### PR TITLE
posix: Pass ucontext to signal handlers

### PR DESCRIPTION
**Pull request by Qwinci**
Change signal code to pass ucontext to signal handlers and add a kHelRegsSignal registerset that matches the linux layout to helLoadRegisters/helStoreRegisters/helQueryRegisterInfo.

### CODE REVIEW

File: kernel/thor/generic/hel.cpp [hunk 1]
```diff
@@ -2020,6 +2060,46 @@ HelError helStoreRegisters(HelHandle handle, int set, const void *image) {
...
+		thread->_executor.general()->rsi = regs[9];
+		thread->_executor.general()->rbp = regs[10];
+		thread->_executor.general()->rbx = regs[11];
+		thread->_executor.general()->rdx = regs[12];
+		thread->_executor.general()->rax = regs[13];
+		thread->_executor.general()->rcx = regs[14];
+		thread->_executor.general()->rsp = regs[15];
+		thread->_executor.general()->rip = regs[16];
+		thread->_executor.general()->rflags = regs[17];
+		thread->_executor.general()->cs = regs[18] | 3;
```
Comment (avdgrinten) [inline kernel/thor/generic/hel.cpp:2090]: `rflags` probably needs to be filtered (e.g., to disallow IOPL changes). We should probably have a mask of allowed bits and keep all other bits invariant.

Restricting `cs` like this sounds fine. Maybe we should add a comment here to explain why it's done though.

Comment (Qwinci) [reply kernel/thor/generic/hel.cpp:2090]: Yeah those are good points. Ill do that and then see if aarch64 needs a similar thing.

File: kernel/thor/generic/hel.cpp [hunk 2]
```diff
@@ -2020,6 +2060,56 @@ HelError helStoreRegisters(HelHandle handle, int set, const void *image) {
...
+		if(!readUserArray(reinterpret_cast<const uintptr_t *>(image), regs, 35))
+			return kHelErrFault;
+		thread->_executor.general()->far = regs[0];
+		for (int i = 0; i < 31; i++)
+			thread->_executor.general()->x[i] = regs[1 + i];
+		thread->_executor.general()->sp = regs[32];
+		thread->_executor.general()->elr = regs[33];
+
+		// Allow N, Z, C, V and SS modifications.
+		constexpr uintptr_t allowedFlagsMask = 0b1111 << 28 | 1 << 21;
```
Comment (qookei) [inline kernel/thor/generic/hel.cpp:2108]: `0b1111 << 28` leads to sign extension, so the mask ends up being `0xFFFFFFFFF0200000`. Making the `0b1111` (e.g. `0b1111U`) unsigned fixes that.

File: posix/subsystem/src/process.cpp [hunk 3]
```diff
@@ -579,9 +579,8 @@ async::result<SignalItem *> SignalContext::fetchSignal(uint64_t mask, bool nonBl
 // This is our signal frame, similar to Linux' struct rt_sigframe.
 #if defined(__x86_64__)
 struct SignalFrame {
-	uint64_t returnAddress; // Address for 'ret' instruction.
-	uintptr_t gprs[kHelNumGprs];
-	uintptr_t pcrs[2];
+	uintptr_t returnAddress;
+	ucontext_t ucontext;
```
Comment (qookei) [inline posix/subsystem/src/process.cpp:583]: This change should also apply to the aarch64 definition below?

**Review by qookei — CHANGES_REQUESTED** [review]
The aarch64 portion is borked

**Review by qookei — COMMENTED** [review]
With the exception of that one comment it looks good now.

File: posix/subsystem/src/process.cpp [hunk 4]
```diff
@@ -707,6 +717,11 @@ async::result<void> SignalContext::raiseContext(SignalItem *item, Process *proce
 	// Store the current register stack on the stack.
 	assert(alignof(SignalFrame) == 8);
 	auto frame = alignFrame(totalFrameSize);
+
+#if defined(__x86_64__)
+	sf.ucontext.uc_mcontext.fpregs = (struct _fpstate*)(frame + sizeof(SignalFrame));
```
[the remark this answers is not in the document]
Comment (Qwinci) [reply posix/subsystem/src/process.cpp:722]: Yeah, I added a todo for aarch64 because I am not sure how it should be implemented.

**Review by qookei — APPROVED** [review]

**Comment by qookei** [comment]
Does not compile on AArch64:
```
../../../src/managarm/posix/subsystem/src/process.cpp:749:26: error: no member named 'x' in 'sigcontext'
        sf.ucontext.uc_mcontext.x[0] = item->signalNumber;
        ~~~~~~~~~~~~~~~~~~~~~~~ ^
../../../src/managarm/posix/subsystem/src/process.cpp:750:26: error: no member named 'x' in 'sigcontext'
        sf.ucontext.uc_mcontext.x[1] = frame + offsetof(SignalFrame, info);
        ~~~~~~~~~~~~~~~~~~~~~~~ ^
../../../src/managarm/posix/subsystem/src/process.cpp:751:26: error: no member named 'x' in 'sigcontext'
        sf.ucontext.uc_mcontext.x[2] = frame + offsetof(SignalFrame, ucontext);
        ~~~~~~~~~~~~~~~~~~~~~~~ ^
../../../src/managarm/posix/subsystem/src/process.cpp:754:26: error: no member named 'x' in 'sigcontext'
        sf.ucontext.uc_mcontext.x[30] = handler.restorerIp;
        ~~~~~~~~~~~~~~~~~~~~~~~ ^
```
AFAICS the array is called `regs`

**Comment by qookei** [comment]
Also, I just noticed that the AArch64 code doesn't actually set PC or SP, or store the modified registers back to set parameters up.

File: posix/subsystem/src/process.cpp [hunk 5]
```diff
@@ -719,25 +735,27 @@ async::result<void> SignalContext::raiseContext(SignalItem *item, Process *proce
...
 #elif defined(__aarch64__)
-	sf.gprs[kHelRegX0] = item->signalNumber;
-	sf.gprs[kHelRegX1] = frame + offsetof(SignalFrame, info);
+	sf.ucontext.uc_mcontext.x[0] = item->signalNumber;
+	sf.ucontext.uc_mcontext.x[1] = frame + offsetof(SignalFrame, info);
+	sf.ucontext.uc_mcontext.x[2] = frame + offsetof(SignalFrame, ucontext);
 
 	// Return address for the 'ret' instruction
-	sf.gprs[kHelRegX30] = handler.restorerIp;
+	sf.ucontext.uc_mcontext.x[30] = handler.restorerIp;
```
Comment (qookei) [inline posix/subsystem/src/process.cpp:754]:
```suggestion
	sf.ucontext.uc_mcontext.x[30] = handler.restorerIp;
	
	sf.ucontext.uc_mcontext.pc = handler.handlerIp;
	sf.ucontext.uc_mcontext.sp = frame;

	HEL_CHECK(helStoreRegisters(thread.getHandle(), kHelRegsSignal, &sf.ucontext.uc_mcontext));
```

Comment (qookei) [inline posix/subsystem/src/process.cpp:754]:
```suggestion
	sf.ucontext.uc_mcontext.regs[30] = handler.restorerIp;

	sf.ucontext.uc_mcontext.pc = handler.handlerIp;
	sf.ucontext.uc_mcontext.sp = frame;

	HEL_CHECK(helStoreRegisters(thread.getHandle(), kHelRegsSignal, &sf.ucontext.uc_mcontext));
```

File: posix/subsystem/src/process.cpp [hunk 6]
```diff
@@ -719,25 +735,27 @@ async::result<void> SignalContext::raiseContext(SignalItem *item, Process *proce
...
+	sf.ucontext.uc_mcontext.gregs[REG_RIP] = handler.handlerIp;
+	sf.ucontext.uc_mcontext.gregs[REG_RSP] = frame;
+
+	HEL_CHECK(helStoreRegisters(thread.getHandle(), kHelRegsSignal, &sf.ucontext.uc_mcontext.gregs));
 #elif defined(__aarch64__)
-	sf.gprs[kHelRegX0] = item->signalNumber;
-	sf.gprs[kHelRegX1] = frame + offsetof(SignalFrame, info);
+	sf.ucontext.uc_mcontext.x[0] = item->signalNumber;
+	sf.ucontext.uc_mcontext.x[1] = frame + offsetof(SignalFrame, info);
+	sf.ucontext.uc_mcontext.x[2] = frame + offsetof(SignalFrame, ucontext);
```
Comment (qookei) [inline posix/subsystem/src/process.cpp:751]:
```suggestion
	sf.ucontext.uc_mcontext.regs[0] = item->signalNumber;
	sf.ucontext.uc_mcontext.regs[1] = frame + offsetof(SignalFrame, info);
	sf.ucontext.uc_mcontext.regs[2] = frame + offsetof(SignalFrame, ucontext);
```

**Comment by Qwinci** [comment]
Should be fixed now, not sure where I got the `x` from.

**Review by qookei — APPROVED** [review]

**Pull request merged**